### ISSUE TITLE
load composite causes strange scene object behaviour

### ISSUE THREAD

**Issue by adybbroe** [pytroll/satpy#45]
Case: First loading a composite, then loading a channel. Calling load on the scene object with the same channel again, satpy reads the channel from the file even though it was already read once. Also, if you have compute=True (default for load method) when loading the channel, also the composite is recomputed!?

Also, the unload didn't seem to have any effect.

Example work flow:

scn = Scene(...)
scn.load(composite_name)
scn.load([0.6])
scn.load([0.6])
scn.unload()
print(scn.datasets.keys())
scn.show(composite_name)

**Comment by davidh-ssec** [comment]
@adybbroe So the issues seem to be:

1. Loading a channel multiple times loads it from the file each time. This may require loading a composite first?
2. Loading a channel after loading a composite recomputes the composite.

My questions for you:

1. What do you expect `scn.unload()` to do that it isn't doing?
2. Do you know if the channel loaded has to be a dependency of the composite for this to fail?
3. Is the composite successfully generated after the first load (no incompatible error warnings)?

**Comment by davidh-ssec** [comment]
@adybbroe In addition to the questions above, could you try out the `develop` branch and see if these behaviors were fixed.

**Comment by adybbroe** [comment]
@davidh-ssec So, thanks, I am testing the develop branch now. It works better at least now. I am not sure it is doing what it should, and at least it doesn't exactly do what I would expect. First to your questions:
1) I expected `scn.unload()` to provide a clean sheet, so that the following code would fail for instance:
```
scn.load([0.6])
scn.unload()
scn.show(0.6)
```
But it doesn't. I can unload but I will still get a nice image displayed back. `scn.unload` doesn't seem to do anything.
2 & 3) No and yes.

**Comment by davidh-ssec** [comment]
@adybbroe The `unload` method is usually automatically called by the `load` method (see `unload` keyword). It is meant to remove any datasets from the Scene that were only needed as prerequisites to create a composite. We could add a `clear` method to delete all datasets from the Scene, but I'm not sure it has much of an advantage compared to just creating a new Scene object.

**Comment by adybbroe** [comment]
@davidh-ssec Ok, I see. No, I don't have a use case in mind where a `clear` method would add a lot. Concerning the `unload` method it is not clear from the docstring at least:
```
In [8]: scn.unload?
Signature: scn.unload(keepables=None)
Docstring:
Unload all loaded composites.
```

**Comment by davidh-ssec** [comment]
Wow, yeah that's not so good.

I think there might be one more case that I didn't handle for reloading or recomputing composites. If a composite (comp2) depends on another composite (comp1) then this would compute comp1 twice.

    scn.load(['comp1'])
    scn.load(['comp2'])

I'll need to check this out and the docstring before closing this.

**Comment by adybbroe** [comment]
@davidh-ssec One more behaviour which I think is a bug, sorry.
So, loading twice seem to be solved now, except when a composite is involved, at least this special one:
```
composite_name = 'day_microphysics'
scn.load([composite_name])
scn.load([10.8])
```
It complains after loading the 10.8 channel that:
`Missing prerequisite for 'DatasetID(name='IR_039...`
And, if instead loading all the channels needed for this composite:
`scn.load([0.6, 3.9, 13.4, 10.8])`
It loads all these channels and derives the reflective part of 3.9 as needed for the composite. Saying `Getting reflective part of IR_039...`
And it keeps calculating this if I try loading all these bands again and again.

**Comment by davidh-ssec** [comment]
@adybbroe Thanks I'll take a look. I don't think this is the case, but it is expected that if you load `composite_name` and you then load one of it's dependencies it *will* require reloading that channel from the file.

**Comment by davidh-ssec** [comment]
@adybbroe I think I've figured this out and it brought out another issue. It may be something that @mraspaud need to discuss further.

What is happening is that when you load the composite it fills a "dependency tree" graph with all of the prerequisites that will be required to make that composite. When you load `10.8`, instead of regenerating all of the dependency tree it first checks if it knows how to resolve `10.8` already. The first issue you're experiencing is that it is taking on "trunk" nodes of the dependency tree and trying to regenerate them. The second issue I've found is that if you were to load `3.7`, the Scene would find the modified version of `3.7` (nir_reflectance) before finding the correct dataset from the reader. You haven't encountered this issue directly yet, but it's still an issue.

For example the dependency tree (just names, not data) looks like this:

    - day_microphysics
        - 0.85
        - 3.7 (nir_reflectance)
            - 3.7
            - 11
            - solar_zenith_angle (optional)
            - 13.4 (optional)
        - 10.8
        
This dependency tree is created when you first load `day_microphysics`. The Scene will attempt to load all of the datasets and then generate the composite. After the composite is generated all of the dependencies are unloaded and only the composite Dataset remains, but the dependency tree still exists.

When you then ask for 10.8 it gathers all "trunk" nodes and decides tries to compute them if they are not in the loaded datasets. So that means it tries to find `3.7 (nir_reflectance)` and `day_microphysics`. The `day_microphysics` composite already exists so it doesn't need to load that. However, the `3.7 (nir_reflectance)` does get computed and that's why you get the warnings about missing prereqs. I think I have a fix for this in the same way I fixed the multi-loading from before.

The issue of asking for `3.7` but getting `3.7 (nir_reflectance)` is something @mraspaud and I will have to discuss. It may be a simple fix, it may not be. I'll let you know when I get these fixed.
